Adding the functionality to copy the symbology raster and SVG to a relative given path

SeqLaz opened the pull request:
Task [QF-1844](https://app.clickup.com/t/2192114/QF-1844)

gounux commented on libqfieldsync/offline_converter.py:
```diff
@@ -280,6 +284,8 @@ def _convert(self, project: QgsProject) -> None:
             elif layer_action == SyncAction.REMOVE:
                 project.removeMapLayer(layer)
 
+            update_symbol_to_project_assets(layer, self._export_filename.parent)
```
```suggestion
            update_symbols_to_project_assets(layer, self._export_filename.parent)
```

gounux commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,85 @@ def is_valid_filepath(path: str) -> bool:
+    layer: QgsVectorLayer, new_path: Optional[Path] = None
+) -> None:
+    """
+    Updates the paths of symbols to a relative QGIS project
```
Would it be possible to add the parameters to the docstring as well ?

SeqLaz replied:
Great!.... I will add them!

gounux commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,85 @@ def is_valid_filepath(path: str) -> bool:
+    if new_path is None:
+        project = QgsProject.instance()
+        project_home = project.homePath()
+        new_path = Path(project_home)
+    destination_dir = new_path / "assets"
```
```suggestion
        new_path = Path(project_home)

    destination_dir = new_path / "assets"
```

SeqLaz replied:
I forgot to give some air!

gounux commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,85 @@ def is_valid_filepath(path: str) -> bool:
+            ):
+                source_path = Path(symbol_layer.path())
+
+                try:
+                    source_path.relative_to(new_path)
+                    continue
+                except ValueError:
+                    # Proceed with copying
+                    pass
```
I find this logic hard to understand

SeqLaz replied:
If the SVG or raster image already exists at the destination, there's no need to copy the file. Simply move on to the next symbol loop. If the path method raises a ValueError, it can cause the code to crash. To prevent this, I used a try-except block.https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.relative_to

SeqLaz replied:
But now reviewing more, you are right is hard to understand, do you think we should remove the relative path check?

suricactus replied:
What are the situations where `source_path.relative_to(new_path)` is throwing a `ValueError`?

SeqLaz replied:
When the new path is not part of the of a subs directories of the source path. This way we evaluate if require the symbol to be copied to the assets folder with a relative path on the new path.

suricactus replied:
Ah, sorry, my bad. I read the wrong documentation from the link above. Ignore my comment.

gounux replied:
Thanks for the explanations !

> do you think we should remove the relative path check?

Actually, I'm okay with the logic of the code, it's just the 

```
try:
  do_something()
  continue
except ValueError:
  pass

do_then_something()
```

in a loop that I find hard to understand.. Maybe I'm just too maniac ?

suricactus replied:
Is using `is_relative_to` from `pathlib` a viable option?

SeqLaz replied:
Indeed, is more viable, I didn't read that method!... Apologize!

gounux commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,85 @@ def is_valid_filepath(path: str) -> bool:
+                    )
+
+            elif hasattr(symbol_layer, "subSymbol"):
+                update_symbol_path_assets(
+                    symbol_layer.subSymbol()
+                )  # Recursively for nested symbology
```
```suggestion
                # Recursively for nested symbology
                update_symbol_path_assets(
                    symbol_layer.subSymbol()
                ) 
```

SeqLaz replied:
Better!

gounux reviewed (commented):
Thank you @SeqLaz ! Did not tried it on my local yet, left a few formalism comments.

gounux commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,92 @@ def is_valid_filepath(path: str) -> bool:
+    destination_dir = new_path / "assets"
+    destination_dir.mkdir(parents=True, exist_ok=True)
+
+    def update_symbols_path_assets(symbol):
```
Is there a reason why using an inner function there ? And not a function one level above (same as the `update_symbols_to_project_assets` one) ?

SeqLaz replied:
No preference, just to make it more simpler as a whole function. But I can move it up

SeqLaz commented:
In the last commit I did many improvements (thanks for the suggestions).

- Move the function for `update_symbols_path_assets` outside the function `update_symbols_to_project_assets`.
- Change the try ... except block for an if statement thanks to the `Path.is_relative_to()` method.
- Removing the recursion in the function `update_symbols_path_assets` not needed since the `QgsRuleBasedRenderer` have the method `symbols()` which return a List with all the symbols even nested one.

gounux reviewed (approved):
After having tested, I can confirm that the icons are well copied to the export folder, and that the new project's symbologies are well configured !

suricactus commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,90 @@ def is_valid_filepath(path: str) -> bool:
+    destination_dir.mkdir(parents=True, exist_ok=True)
+
+    for symbol_layer in symbol.symbolLayers():
+        if isinstance(
+            symbol_layer, (QgsSvgMarkerSymbolLayer, QgsRasterMarkerSymbolLayer)
+        ):
+            source_path = Path(symbol_layer.path())
+
+            if not source_path.is_relative_to(new_path):
+                if source_path.exists():
+                    destination_path_file = destination_dir / source_path.name
+
+                    if not destination_path_file.exists():
+                        shutil.copy2(source_path, destination_path_file)
+
+                    symbol_layer.setPath(
+                        str(destination_path_file.relative_to(new_path))
+                    )
```
```suggestion
        # filter out only symbology that includes SVG and skip the rest
        if not isinstance(
            symbol_layer, (QgsSvgMarkerSymbolLayer, QgsRasterMarkerSymbolLayer)
        ):
            continue

        source_path = Path(symbol_layer.path())

        # the symbol is already broken, it's file is not reachable
        if source_path.is_file():
            continue

        # if the symbol's path is already relative, we have nothing to do
        if source_path.is_relative_to(new_path):
            continue

        destination_path_file = destination_dir / source_path.name

        if not destination_path_file.exists():
            shutil.copy2(source_path, destination_path_file)

        symbol_layer.setPath(
            str(destination_path_file.relative_to(new_path))
        )
```
I hope it is a bit morer readable than 8 levels of nesting.

suricactus replied:
I think the whole codebase is prefering the `joinpath` over `/` operator.

suricactus replied:
What I was wondering is since we do a `shutil.copy2`, there might be something in `assets` already. Is there any chance to embed the SVG symbol dynamically, or it is way too complicated?

SeqLaz replied:
Great!... Indeed is more readable!.... Thanks @suricactus!

suricactus commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,90 @@ def is_valid_filepath(path: str) -> bool:
+        categories = renderer.categories()
+        if categories:
+            for index in range(len(categories)):
+                # Get a fresh category object each time,
+                # if not the new symbol path is not updated.
```
Good comment, but I beleive we can rewrite for clarity what it exactly means.

SeqLaz replied:
Sure will!

suricactus commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,90 @@ def is_valid_filepath(path: str) -> bool:
+    elif isinstance(renderer, QgsRuleBasedRenderer):
+        for rule in renderer.rootRule().children():
+            symbols = rule.symbols()
+            if symbols:
+                for symbol in symbols:
+                    update_symbols_path_assets(symbol, new_path)
```
```suggestion
            # ???
            if not symbols:
                continue

            for symbol in symbols:
                update_symbols_path_assets(symbol, new_path)
```

SeqLaz replied:
Great!...Thanks @suricactus!

SeqLaz commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +226,90 @@ def is_valid_filepath(path: str) -> bool:
+                    destination_path_file = destination_dir / source_path.name
+
+                    if not destination_path_file.exists():
+                        shutil.copy2(source_path, destination_path_file)
```
Let me give it a try!

SeqLaz replied:
I successfully update the code for embed the markers on the project, check the new video on the description.

SeqLaz commented:
Hey @suricactus and @gounux in the last commit I added all the suggestions, thanks for it.

Also changed the logic from copy the files to assets to embed the symbology in the projects. See the new video in the description.

suricactus commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +227,82 @@ def is_valid_filepath(path: str) -> bool:
+
+        with open(source_path, "rb") as file:
+            file_data = file.read()
+            encoded_data = base64.b64encode(file_data).decode()
+            symbol_layer.setPath(f"base64:{encoded_data}")
```
Isn't there more QGIS way to achieve the same? I don't know, just asking.

SeqLaz replied:
Unfortunately as method there is not a way to set embedded markers. Only the setPath()
https://qgis.org/pyqgis/3.40/core/QgsRasterMarkerSymbolLayer.html#qgis.core.QgsRasterMarkerSymbolLayer.setPath

By making the embed symbol using the layer style in QGIS UI the result is the same as programmatically using the method setPath(f"base64:{encoded_data}")

suricactus replied:
Ok, we can live with it then.

gounux commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +227,82 @@ def is_valid_filepath(path: str) -> bool:
+
+def embed_layer_symbols_on_project(layer: QgsVectorLayer) -> None:
+    """
+    Updates the paths of symbols to a relative QGIS project
```
Is this function description still accurate, now that symbols are embedded ?

SeqLaz replied:
Ouch, you are right!..... I forgot to change it!

gounux commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +227,82 @@ def is_valid_filepath(path: str) -> bool:
+    if not renderer:
+        return
+
+    if isinstance(renderer, QgsSingleSymbolRenderer):
+        symbol = renderer.symbol()
+        if symbol:
+            update_symbols_to_embedded(symbol=symbol)
+
+    elif isinstance(renderer, QgsRuleBasedRenderer):
+        for rule in renderer.rootRule().children():
+            symbols = rule.symbols()
+
+            if not symbols:
+                continue
+
+            for symbol in symbols:
+                update_symbols_to_embedded(symbol=symbol)
+
+    elif isinstance(renderer, QgsCategorizedSymbolRenderer):
```
Naive question (I don't have the answer): could we miss some [QgsFeatureRenderer](https://qgis.org/pyqgis/3.40/core/QgsFeatureRenderer.html) sub-class implementations ?

SeqLaz replied:
I think the renderers that allows the SVG and Rasters markers are QgsSingleSymbolRenderer, QgsRuleBasedRenderer, QgsCategorizedSymbolRenderer, as per the [QgsEmbeddedSymbolRenderer](https://qgis.org/pyqgis/3.40/core/QgsEmbeddedSymbolRenderer.html) this information should be directly in the feature.

But I think is better to confirm with @suricactus

suricactus replied:
I don't know, but I think supporting `QgsFeatureRenderer` is already a big deal. Let us the community test it and address it in the future.

gounux reviewed (commented):
Had the opportunity to test again, indeed the symbols are now embedded in the QGIS project for QField!

![image](https://github.com/user-attachments/assets/86aff4a5-8ed5-4722-9a15-3a9757d967bc)

Thanks @SeqLaz, nice work!

suricactus commented:
Thanks!

suricactus commented on libqfieldsync/utils/file_utils.py:
```diff
@@ -216,3 +227,82 @@ def is_valid_filepath(path: str) -> bool:
+        if not source_path.is_file():
+            continue
+
+        with open(source_path, "rb") as file:
```
@SeqLaz I just realized that you have removed the check for a relative path, I believe we should keep it.

If we have this structure:
```
- project.qgs
- asset.svg
```
It should actually use the asset and not force embedding.

SeqLaz replied:
It should be in this PR https://github.com/opengisch/libqfieldsync/pull/108
